src/Compilers/CSharp/Portable/Utilities/InterceptableLocation.cs

```diff
@@ -0,0 +1,183 @@
+
+    private readonly ImmutableArray<byte> _checksum;
+    private readonly string _path;
+    private readonly int _position;
+    private readonly int _lineNumberOneIndexed;
+    private readonly int _characterNumberOneIndexed;
+    private string? _lazyData;
+
+    internal InterceptableLocation1(ImmutableArray<byte> checksum, string path, int position, int lineNumberOneIndexed, int characterNumberOneIndexed)
+    {
+        Debug.Assert(checksum.Length == ContentHashLength);
+        Debug.Assert(path is not null);
+        Debug.Assert(position >= 0);
+        Debug.Assert(lineNumberOneIndexed > 0);
+        Debug.Assert(characterNumberOneIndexed > 0);
+
+        _checksum = checksum;
+        _path = path;
+        _position = position;
+        _lineNumberOneIndexed = lineNumberOneIndexed;
+        _characterNumberOneIndexed = characterNumberOneIndexed;
+    }
+
+    public override string GetDisplayLocation()
+    {
+        // e.g. `C:\project\src\Program.cs(12,34)`
+        return $"{_path}({_lineNumberOneIndexed},{_characterNumberOneIndexed})";
+    }
+
+    public override string ToString() => GetDisplayLocation();
+
+    public override int Version => 1;
+    public override string Data
+    {
+        get
+        {
+            if (_lazyData is null)
+                _lazyData = makeData();
+
+            return _lazyData;
+
+            string makeData()
+            {
+                var builder = PooledBlobBuilder.GetInstance();
+                builder.WriteBytes(_checksum, start: 0, 16);
+                builder.WriteInt32(_position);
+
+                var displayFileName = Path.GetFileName(_path);
+                builder.WriteUTF8(displayFileName);
+
+                var bytes = builder.ToArray();
+                builder.Free();
+                return Convert.ToBase64String(bytes);
+            }
+        }
+    }
+
+    internal static (ReadOnlyMemory<byte> checksum, int position, string displayFileName)? Decode(string? data, Location diagnosticLocation, BindingDiagnosticBag diagnostics)
+    {
+        if (data is null)
+        {
+            diagnostics.Add(ErrorCode.ERR_InterceptsLocationDataInvalidFormat, diagnosticLocation);
+            return null;
+        }
+
+        byte[] bytes;
+        try
+        {
+            bytes = Convert.FromBase64String(data);
+        }
+        catch (FormatException)
+        {
+            diagnostics.Add(ErrorCode.ERR_InterceptsLocationDataInvalidFormat, diagnosticLocation);
+            return null;
+        }
+
+        // format:
+        // - 16 bytes of target file content hash (xxHash128)
+        // - int32 position (little endian)
+        // - utf-8 display filename
+        const int hashIndex = 0;
+        const int hashSize = 16;
+        const int positionIndex = hashIndex + hashSize;
+        const int positionSize = sizeof(int);
+        const int displayNameIndex = positionIndex + positionSize;
+        const int minLength = displayNameIndex;
```

Comment:
>minLength [](http://example.com/codeflow?start=18&length=9)

Consider inlining, to reduce the number of constants (slightly).

Reply:
I prefer this factoring as it's more self-documenting.